openage/convert/interface/visgrep.py

```diff
@@ -0,0 +1,389 @@
+
+
+def img_pixel_get(img, point):
+    """
+    Get pixel color or zero if outside bounds.
+    """
+    if point.x < img.size[0] and point.y < img.size[1]:
+        return Pixel(*img.getpixel(point))
+    else:
+        return Pixel(0, 0, 0, 0)
+
+
+def img_subimage_find(master, find, start_from, tolerance, find_next):
+    """
+    Fuzzily find a part of an image that matches the patern.
+    """
+    x_end = master.size[0] - find.size[0]
+    y_end = master.size[1] - find.size[1]
+
+    if find_next:
+        start_from = Point(start_from.x + 1, start_from.y)
+
+    logging.debug("Starting from %d", start_from)
+    logging.debug("End %d,%d", x_end, y_end)
+
+    # Loop the whole freakin image looking for this sub image, but not past edges
+    for y_it in range(start_from.y, y_end + 1):
+        logging.debug("Begin subimg find loop for y %d of %d", y_it, y_end)
+        for x_it in range(start_from.x, x_end + 1):
+            point = Point(x_it, y_it)
+            logging.debug("Begin subimg find loop for x: %d,%d", point.x, point.y)
+            badness = img_subimage_cmp(master, find, point, tolerance)
+            if badness <= tolerance:
+                logging.debug("Found subimage at %d,%d", point.x, point.y)
+                return FoundResult(badness, point)
+
+    # No match
+    return FoundResult(-1, Point(-1, -1))
+
+
+def img_pixel_cmp(pix, other_pix):
+    """
+    o is the compare from pixel, assumed to be from a pattern. It's transparency
+    is the transparency used to modify the tolerance value
+      return( memcmp( &p, &o, sizeof( PIXEL ) ) );
+    make tolerance mean something
+    """
+    difference = abs(pix.r - other_pix.r) + abs(pix.g - other_pix.g) + abs(pix.b - other_pix.b)
+    transparentness = other_pix.a
+    difference = difference * (transparentness / 255)
+    logging.debug("Difference: %d", difference)
+    return difference
+
+
+def img_subimage_cmp(master, subimage, where, tolerance):
+    """
+    Returns 0 if subimage is inside master at where, like *cmp usually does for other stuff
+    otherwise returns an integer of how different the match is, for each color component
+    value off.  tolerance is how high to go before bailing.  set lower to avoid processing
+    lots of extra pixels, it will just ret when tolerance is met
+    """
+    mpt = Point(0, 0)
+
+    logging.debug("Comparing subimage where=%d,%d", where.x, where.y)
+
+    # Check if subimage even fits in masterimage at POINT
+    if ((where.x + subimage.size[0]) > master.size[0] or
+            (where.y + subimage.size[1]) > master.size[1]):
+        # Superbad
+        logging.debug("Subimage would not fit here")
+        return 1000
+
+    badness = 0
+    for sptx in range(subimage.size[0]):
+        for spty in range(subimage.size[1]):
+            # Map U/V to X/Y
+            mpt = Point(sptx + where.x, spty + where.y)
+
+            logging.debug("Pixel cmp @ main: %d,%d, subimage: %d,%d", mpt.x, mpt.y, sptx, spty)
```

Comment:
maybe you could think about using our internal logger, but then the script would depend on libopenage and can't work standalone, so leaving it like this is probably best.